### ISSUE TITLE
Problems syncing local name changes and deletions to the server

### ISSUE THREAD

**Issue by adrianghc** [abraunegg/onedrive#195]
### Bug Report Details ###
**Describe the bug**
When deleting a file or directory on the local copy, this is not synced up to OneDrive. When renaming a file or directory on the local copy, a new copy is synced up to OneDrive and a copy with the old name is kept. All other scenarios work in both directions, as far as I can tell.

**Application and Operating System Details:**
- Linux stuart 4.9.0-7-armmp-lpae #1 SMP Debian 4.9.110-3+deb9u2 (2018-08-13) armv7l GNU/Linux
- I'm running a headless client
- The client version is onedrive v2.1.4
- I use a personal Microsoft account for OneDrive
- I compiled with ldc2-1.10.0-linux-armhf

**To Reproduce**
To reproduce, create a new file or folder and sync with _./onedrive --synchronize_. The result is that nothing happens. The console output shows no errors and is simply

> Initializing the Synchronization Engine ...
> Syncing changes from OneDrive …

When renaming an existing file that has already been uploaded to OneDrive, e.g. renaming _test.txt_ to _test2.txt_, _test.txt_ is kept on the server and _test2.txt_ is uploaded additionally. The console output in that case is

> Initializing the Synchronization Engine ...
> Syncing changes from OneDrive 

**Additional context**
I only sync a single directory that I configured in sync_list. Everything detailed here happens within that directory.

I can provide further details as needed.

**Comment by abraunegg** [comment]
@adrianghc

Can you advise if you did an upgrade from the skilion version?

I know that this issue manifests itself if you are upgrading and you do not delete the items.* files from your ~/.config/ondrive directory

**Comment by adrianghc** [comment]
>Can you advise if you did an upgrade from the skilion version?
I know that this issue manifests itself if you are upgrading and you do not delete the items.* files from your ~/.config/ondrive directory

@abraunegg I did an upgrade, yes, but I did a _--resync_, which I thought would reset the database. Which files do you suggest I delete?

**Comment by abraunegg** [comment]
> I only sync a single directory that I configured in sync_list

Can you remove using sync_list and specify the local directory to sync by using the --single-directory option?

What is the full output of what you have configured in sync_list ?

**Comment by adrianghc** [comment]
> Can you remove using sync_list and specify the local directory to sync by using the --single-directory option?

The client seems to get stuck at
>Syncing changes from selected OneDrive path …

doing this.

> What is the full output of what you have configured in sync_list ?

```
Backups

```

> delete the items.* files from your ~/.config/ondrive directory

I only see _items.sqlite3_, are there supposed to be others? Deleting it did not fix this.

**Comment by abraunegg** [comment]
@adrianghc 
Can you please do the following:

1. Ensure that all 'onedrive' processes are stopped - ensure that any process you are running is not the skilion version 
2. Remove `~/.config/ondrive/items.*`
3. Remove any configuration in sync_list (delete the file)
4. Use the following command: `onedrive --synchronize --verbose --single-directory 'Backups'`
5. Post full output from the above command in step 3 above here

**Comment by adrianghc** [comment]
@abraunegg It's running now and taking a while because I have many directories. Would you like the output only after first execution or also after syncing item deletions?

**Comment by abraunegg** [comment]
@adrianghc 
No need to show log output if it is working using the suggested command.

What we need to work out is what is/was causing your issue:

* Bad upgrade / bad local database file due to 'skilion' version
* Incorrect processing & handling of 'sync_list'

**Comment by adrianghc** [comment]
Well, I already tried deleting the database with no effect, and I also tried with the _-single-directory_ flag but it takes much longer than without the flag. I should add that the _Backups_ directory is empty so what it's taking so long with is the stuff in all the other folders that it's skipping (right now it's printing a lot of _Remote change discarded - not in --single-directory scope_). Is it normal for sync to take much longer with the _-single-directory_ flag than when you specify a _sync_list_?

**Comment by abraunegg** [comment]
@adrianghc 
Can you please be more articulate in your 'use case' here so I can help. What is it your trying to achieve?

> Remote change discarded - not in --single-directory scope

When this is printing, if is discarding changes you made online, on your OneDrive account, outside of the path of 'Backup'

> Is it normal for sync to take much longer with the -single-directory flag than when you specify a sync_list?

sync_list operates totally differently to --single-directory
* sync_list looks at files & folders that = each line item in the sync_list file
* --single-directory re-points the root logically to look at the specified path to sync all changes only in that path locally and from OneDrive

Now given that this is a 'Backup' folder - I am going to assume here that your doing something locally, and wanting to 'backup' to OneDrive the contents of this folder.

If this is the case, I would also use the '--upload-only' flag so that you are only uploading your local changes (new files / deletes / renames) to OneDrive, and ignoring any changes that you make on OneDrive.

**Comment by adrianghc** [comment]
Right now I'm running ```onedrive --synchronize --verbose --single-directory 'Backups'``` as you suggested after performing steps 1 to 4. It's taking very long - previously I would have assumed that it's because it's going through all the directories in my OneDrive account and discarding all those that are not in _Backups_. However, your description of what _--single-directory_ does doesn't fit with this, so I guess I don't understand what's happening and what's taking so long. Right now it's still printing _Remote change discarded - not in --single-directory scope_ every few seconds.

>  I am going to assume here that your doing something locally, and wanting to 'backup' to OneDrive the contents of this folder.

Actually, I don't want to backup the contents of the folder but the contents of the folder _are_ backups of something else.

> If this is the case, I would also use the '--upload-only' flag so that you are only uploading your local changes (new files / deletes / renames) to OneDrive, and ignoring any changes that you make on OneDrive.

I forgot to mention that I also tried the _--upload-only_ flag but the problem persists.

**Comment by abraunegg** [comment]
This should be the command you use:
`onedrive --synchronize --verbose --single-directory 'Backups'`

This option will do a bi-directional sync, and as you have seen, check all the changes as provided by OneDrive for matching the 'Backups' directory scope before applying the changes locally. Once you are happy with the operation, drop the verbosity switch.

Depending if you are then only uploading changes from your local system to OneDrive, add `--upload-only` to the above command. This will not check OneDrive for changes, thus not process any remote changes to your local system.

**Comment by abraunegg** [comment]
**Examples:**
Sync of 'Backups' Folder:
```
[user@centos7full onedrive]$ onedrive --synchronize --verbose --single-directory 'Backups'
Loading config ...
Using Config Dir: /home/user/.config/onedrive
No config file found, using defaults
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /home/user/OneDrive
Initializing the Synchronization Engine ...
Account Type: personal
Default Drive ID: <redacted>
Default Root ID: <redacted>
Remaining Free Space: 5368689790
Fetching details for OneDrive Root
OneDrive Root exists in the database
Syncing changes from this selected path: Backups
Syncing changes from selected OneDrive path ...
Getting path details from OneDrive ...
Applying changes of Path ID: <redacted>
Uploading differences of Backups
Processing Backups
The directory has not changed
Uploading new items of Backups
Getting path details from OneDrive ...
Applying changes of Path ID: <redacted>
```

Add a file to 'Backups' Folder & Sync:
```
[user@centos7full onedrive]$ onedrive --synchronize --verbose --single-directory 'Backups'
Loading config ...
Using Config Dir: /home/user/.config/onedrive
No config file found, using defaults
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /home/user/OneDrive
Initializing the Synchronization Engine ...
Account Type: personal
Default Drive ID: <redacted>
Default Root ID: <redacted>
Remaining Free Space: 5368689790
Fetching details for OneDrive Root
OneDrive Root exists in the database
Syncing changes from this selected path: Backups
Syncing changes from selected OneDrive path ...
Getting path details from OneDrive ...
Applying changes of Path ID: <redacted>
Uploading differences of Backups
Processing Backups
The directory has not changed
Uploading new items of Backups
Uploading file Backups/file1 ... done.
Remaining free space: 5368689790
Getting path details from OneDrive ...
Applying changes of Path ID: <redacted>
```

Move file1 to file2 locally:
```
[user@centos7full onedrive]$ onedrive --synchronize --verbose --single-directory 'Backups'
Loading config ...
Using Config Dir: /home/user/.config/onedrive
No config file found, using defaults
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /home/user/OneDrive
Initializing the Synchronization Engine ...
Account Type: personal
Default Drive ID: <redacted>
Default Root ID: <redacted>
Remaining Free Space: 5368689790
Fetching details for OneDrive Root
OneDrive Root exists in the database
Syncing changes from this selected path: Backups
Syncing changes from selected OneDrive path ...
Getting path details from OneDrive ...
Applying changes of Path ID: <redacted>
Uploading differences of Backups
Processing Backups
The directory has not changed
Processing file1
The file has been deleted locally
Deleting item from OneDrive: Backups/file1
Uploading new items of Backups
Uploading file Backups/file2 ... done.
Remaining free space: 5368689790
Getting path details from OneDrive ...
Applying changes of Path ID: <redacted>
```

Rename file online:
```
[user@centos7full onedrive]$ onedrive --synchronize --verbose --single-directory 'Backups'
Loading config ...
Using Config Dir: /home/user/.config/onedrive
No config file found, using defaults
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /home/user/OneDrive
Initializing the Synchronization Engine ...
Account Type: personal
Default Drive ID: <redacted>
Default Root ID: <redacted>
Remaining Free Space: 5368689790
Fetching details for OneDrive Root
OneDrive Root exists in the database
Syncing changes from this selected path: Backups
Syncing changes from selected OneDrive path ...
Getting path details from OneDrive ...
Applying changes of Path ID: <redacted>
Moving Backups/file2 to Backups/file2-rename_online
Uploading differences of Backups
Processing Backups
The directory has not changed
Processing file2-rename_online
The file has not changed
Uploading new items of Backups
Getting path details from OneDrive ...
Applying changes of Path ID: <redacted>
```

**Comment by abraunegg** [comment]
Examples:
```
[user@centos7full onedrive]$ cat ~/.config/onedrive/sync_list 
Backups
[user@centos7full onedrive]$ 
```

Sync of 'Backups' Folder:
```
[user@centos7full onedrive]$ ./onedrive --synchronize --verbose 
Loading config ...
Using Config Dir: /home/user/.config/onedrive
No config file found, using defaults
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /home/user/OneDrive
Initializing the Synchronization Engine ...
Account Type: personal
Default Drive ID: <redacted>
Default Root ID: <redacted>
Remaining Free Space: 5368689790
Fetching details for OneDrive Root
OneDrive Root does not exist in the database. We need to add it.
Added OneDrive Root to the local database
Syncing changes from OneDrive ...
Applying changes of Path ID: <redacted>
Uploading differences of .
Processing root
The directory has not changed
Processing Backups
The directory has not changed
Processing file2-rename_online
The file has not changed
Uploading new items of .
Skipping item - path excluded: ./Document1.docx
Skipping item - path excluded: ./Documents
Skipping item - path excluded: ./4B
Skipping item - path excluded: ./3A
Skipping item - path excluded: ./file
Skipping item - path excluded: ./1
Applying changes of Path ID: <redacted>
```

Make change online of file outside 'Backups' folder:
```
[user@centos7full onedrive]$ ./onedrive --synchronize --verbose
Loading config ...
Using Config Dir: /home/user/.config/onedrive
No config file found, using defaults
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /home/user/OneDrive
Initializing the Synchronization Engine ...
Account Type: personal
Default Drive ID: <redacted>
Default Root ID: <redacted>
Remaining Free Space: 5368689591
Fetching details for OneDrive Root
OneDrive Root exists in the database
Syncing changes from OneDrive ...
Applying changes of Path ID: <redacted>
Uploading differences of .
Processing root
The directory has not changed
Processing Backups
The directory has not changed
Processing file2-rename_online
The file has not changed
Uploading new items of .
Skipping item - path excluded: ./Document1.docx
Skipping item - path excluded: ./Documents
Skipping item - path excluded: ./4B
Skipping item - path excluded: ./3A
Skipping item - path excluded: ./file
Skipping item - path excluded: ./1
Applying changes of Path ID: 66D53BE8A5056ECA!101
```
Add new file locally:
```
[user@centos7full onedrive]$ ./onedrive --synchronize --verbose
Loading config ...
Using Config Dir: /home/user/.config/onedrive
No config file found, using defaults
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /home/user/OneDrive
Initializing the Synchronization Engine ...
Account Type: personal
Default Drive ID: <redacted>
Default Root ID: <redacted>
Remaining Free Space: 5368689591
Fetching details for OneDrive Root
OneDrive Root exists in the database
Syncing changes from OneDrive ...
Applying changes of Path ID: <redacted>
Uploading differences of .
Processing root
The directory has not changed
Processing Backups
The directory has not changed
Processing file2-rename_online
The file has not changed
Uploading new items of .
Skipping item - path excluded: ./Document1.docx
Skipping item - path excluded: ./Documents
Skipping item - path excluded: ./4B
Skipping item - path excluded: ./3A
Uploading file ./Backups/file3 ... done.
Remaining free space: 5368689591
Skipping item - path excluded: ./file
Skipping item - path excluded: ./1
Applying changes of Path ID: <redacted>
```

**Comment by abraunegg** [comment]
**Example showing error described above**
* Use of sync_list & --single-directory together
```
[user@centos7full onedrive]$ onedrive --synchronize --verbose --single-directory 'Backups'
Loading config ...
Using Config Dir: /home/user/.config/onedrive
No config file found, using defaults
Initializing the OneDrive API ...
Opening the item database ...
All operations will be performed in: /home/user/OneDrive
Initializing the Synchronization Engine ...
Account Type: personal
Default Drive ID: <redacted>
Default Root ID: <redacted>
Remaining Free Space: 5368689591
Fetching details for OneDrive Root
OneDrive Root exists in the database
Syncing changes from this selected path: Backups
Syncing changes from selected OneDrive path ...
Getting path details from OneDrive ...
Applying changes of Path ID: <redacted>
Remote change discarded - not in --single-directory scope
Remote change discarded - not in --single-directory scope
Remote change discarded - not in --single-directory scope
Remote change discarded - not in --single-directory scope
Remote change discarded - not in --single-directory scope
Remote change discarded - not in --single-directory scope
Remote change discarded - not in --single-directory scope
Remote change discarded - not in --single-directory scope
Remote change discarded - not in --single-directory scope
Remote change discarded - not in --single-directory scope
Remote change discarded - not in --single-directory scope
Uploading differences of Backups
Processing Backups
The directory has not changed
Processing file2-rename_online
The file has not changed
Processing file3
The file has not changed
Processing file4
The file has not changed
Uploading new items of Backups
Getting path details from OneDrive ...
Applying changes of Path ID: <redacted>
```

**Comment by adrianghc** [comment]
It seems that using ```onedrive --synchronize --verbose --single-directory 'Backups'``` without a _sync_list_ file indeed works (and it's also much quicker after the second time). So I suppose that is the way to go, although it doesn't seem quite right to me that it wouldn't work using the _sync_list_ instead of the _--single-directory_, unless I'm still misunderstanding something. Anyway, I'll just use the flag from now on. Thank you for your help.

**Comment by abraunegg** [comment]
The most plausible explanation here is as to why you had problems:
* database issues / upgrade issues from 'skilion' version
* misconfiguration in sync_list - space at the end of the line, unwarranted new line at end of file
* attempting to use sync_list & --single-directory together

>  Also, you can't use the OneDrive service (as in Linux service) with the flag, or can you?
Yes you can - you just edit the service file with the right options.

**Comment by adrianghc** [comment]
Number 1 isn't plausible because I cleared the database. Number 3 isn't plausible either because I did try sync_list without the flag. So I guess this leaves us with number 2, where I did indeed have an empty line at the end of the file - why would that be the problem, though? After all, other operations in the folder 'Backups' were properly recognized.

> Yes you can - you just edit the service file with the right options.
I see, thank you very much. I don't have prior experience with Linux services so that didn't occur to me.
